[backport v1.14] do some necessary backport for ARC

IRISZZW opened the pull request:
1. fix the bug where regular irq comes out in user mode
2. fix the bug where buffer validate should be atomic
3. add handling for r25, r30
4. Preserve STATUS32 flags while setting AE flag

IRISZZW committed:
arc: Preserve STATUS32 flags while setting AE flag

KFLAG instruction might affect multiple flags in STATUS32 register
and so when we need just AE-bit to be set we need first read current
state of STATUS32, then change our bit and set STATUS32 again.

Otherwise critical flags including stack checking, unaligned access etc
will be dropped for good.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

ruuddw commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -337,6 +337,13 @@ _skip_sys_power_save_idle_exit:
+ */
+	push r25
+	push r30
+	push r58
```
for the accumulator registers (r58/r59), the CONFIG_ARC_HAS_ACCL_REGS should be used/backported as well. Only DSP-flavors of ARC have these.

andrewboie replied:
@ruuddw this is a backport PR, is there a patch in master for this one?

IRISZZW replied:
ARC FPU is not support in 1.14 (supported in PR: #16921), so we can remove the save and restore of this two registers.

ruuddw commented on arch/arc/core/mpu/arc_mpu_v3_internal.h:
```diff
@@ -596,13 +596,15 @@ int arc_core_mpu_buffer_validate(void *addr, size_t size, int write)
+		r_index = -EPERM;
 	}
 
-	return -EPERM;
```
where is the unlock?

IRISZZW replied:
so stupid mistake, fixed.

ruuddw reviewed (changes requested):
Looks like some crucial parts were missed in the backport, see comments.

IRISZZW committed:
arc: add handling for r25, r30

r25 and r30 are caller-saved scratch reg.  so need to
be saved before jumping to interrupt handlers

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW committed:
arc: fix the bug where buffer validate should be atomic

the access to mpu regs in buffer validate should be atomic.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

IRISZZW committed:
arc: fix the bug where regular irq comes out in user mode

if USERSPACE is configured, it needs to record the user/kernel mode
of interrupted thread, because the switch of aux_sec_k_sp/aux_user_sp
depends on the aux_irq_act's U bit.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

ruuddw commented:
Looks ok to me now. @IRISZZW can you confirm regression tests pass, also on hardware boards?
